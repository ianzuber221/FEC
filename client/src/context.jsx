import React, { useContext, useState } from 'react';

const CurrentProductContext = React.createContext();
const UpdateCurrentProductContext = React.createContext();

export function useCurrentProductContext() {
  return useContext(CurrentProductContext);
}

export function useCurrentProductUpdate(obj) {
  return useContext(UpdateCurrentProductContext);
}
export function CurrentProductProvider({ children }) {
  const [currentProduct, setCurrentProduct] = useState('37314');

  function changeCurrentProduct(newProduct) {
    setCurrentProduct(newProduct);
  }

  return (
    <CurrentProductContext.Provider value={currentProduct}>
      <UpdateCurrentProductContext.Provider value={changeCurrentProduct}>
        {children}
      </UpdateCurrentProductContext.Provider>
    </CurrentProductContext.Provider>
  );
}