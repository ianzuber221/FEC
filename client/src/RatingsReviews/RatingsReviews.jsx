import React, { useState, useEffect, forwardRef } from 'react';
import axios from 'axios';
import styled from 'styled-components';
import { parseISO, compareAsc } from 'date-fns';
import Ratings from './Ratings/Ratings';
import Reviews from './Reviews/Reviews';
import getTotalRatings from './lib/getTotalRatings';
import { useCurrentProductContext } from '../context';

const RatingsReviews = forwardRef((props, ref) => {
  const [meta, setMeta] = useState({});
  const [totalRatings, setTotalRatings] = useState(0);
  const [isLoaded, setIsLoaded] = useState(false);
  const [filteredByRating, setFilteredByRating] = useState(false);
  const [reviews, setReviews] = useState([]);
  const [page, setPage] = useState(1);
  const [filtered, setFiltered] = useState([]);
  const [average, setAverage] = useState(0);
  const [ratings, setRatings] = useState([]);
  const [ratingFilter, setRatingFilter] = useState([]);
  const [sortedReviews, setSortedReviews] = useState([]);
  const [sortOption, setSortOption] = useState('');
  const [sorted, setSorted] = useState(false);
  const [sortCount, setSortCount] = useState(0);

  const productId = useCurrentProductContext();

  const getReviews = () => {
    axios.get(`${process.env.API_URL}/reviews?product_id=${productId}&count=500`, {
      headers: {
        Authorization: process.env.AUTH_KEY,
      },
    })
      .then((response) => {
        // console.log('got reviews.')
        setReviews(response.data.results);
      })
      .then(() => {
        if (sorted) {
          sort(sortOption);
        }
      })
      .catch((err) => console.log('error fetching reviews', err));
  };

  const getRatings = () => {
    axios.get(`${process.env.API_URL}/reviews/meta?product_id=${productId}`, {
      headers: {
        Authorization: process.env.AUTH_KEY,
      },
    })
      .then((response) => {
        // console.log('got ratings.')
        const sum = getTotalRatings(response.data.ratings)[0];
        const totalRatings = getTotalRatings(response.data.ratings)[1];
        setMeta(response.data);
        setAverage((sum / totalRatings).toFixed(1));
        setRatings(response.data.ratings);
        setTotalRatings(totalRatings);
        setIsLoaded(true);
      })
      .catch((err) => console.log('error fetching ratings', err));
  };

  // useEffect(() => {
  //   getReviews();
  //   getRatings();
  // }, []);

  useEffect(() => {
    getRatings();
    getReviews();
  }, [productId]);

  const setRatingsFilter = (rating) => {
    const index = ratingFilter.indexOf(rating);
    if (index === -1) {
      setRatingFilter((a) => [...a, rating]);
    } else {
      const temp = ratingFilter;
      temp.splice(index, 1);
      setRatingFilter(() => [...temp]);
    }
  };

  const getByRating = () => {
    // set temp as current list of reviews,
    // filter temp to fit ratings filter,
    // set state reviews to be temp.
    const temp = reviews;
    const arr = ratingFilter;
    const filteredReviews = temp.filter((review) => {
      // if review.rating is found in array, return that review.
      const index = arr.indexOf(`${review.rating}`);
      if (index !== -1) {
        return review;
      }
    });
    setFiltered(filteredReviews);
  };

  useEffect(() => {
    if (ratingFilter.length > 0) { // once ratingfilter updates, is longer than 0
      getByRating();
      setFilteredByRating(true);
    } else if (isLoaded) { // if its not longer than 0, empty array, just get regular reviews.
      getReviews();
      setFilteredByRating(false);
      setFiltered([]);
    }
  }, [ratingFilter]);

  const sort = (sortMethod) => {
    // setSorted(false);
    const temp = reviews;
    let list = document.getElementById('reviewlist1');
    list.scrollTo(0, 0);
    setPage(1);
    if (sortMethod === 'helpful') {
      temp.sort((a, b) => b.helpfulness - a.helpfulness);
      setSortedReviews(() => temp);
      setSortOption(sortMethod);
      setSortCount(sortCount + 1); // trigger effect
      setSorted(true);
    }
    if (sortMethod === 'newest') {
      temp.sort((a, b) => compareAsc(parseISO(b.date), parseISO(a.date)));
      setSortedReviews(() => temp);
      setSortOption(sortMethod);
      setSortCount(sortCount + 1); // trigger effect
      setSorted(true);
    }
    if (sortMethod === 'relevance') { // relevance, just reset sort and get og reviews
      setSorted(false);
      setSortedReviews([]);
      getReviews();
    }
  };

  useEffect(() => {
    // listen to sortoption, if sortoption changes, reset page to 1.
    setPage(1);
  }, [sortOption]);

  useEffect(() => {
    if (sorted) {
      setReviews[sortedReviews];
    }
  }, [sortCount]);

  const moreReviews = () => {
    setPage(page + 1);
  };

  const scrollMore = () => { // only works when its not filtered by rating.
    // if (!filteredByRating) {
    setPage(page + 1);
    // }
  };

  const onFormSubmit = () => {
    let list = document.getElementById('reviewlist1');
    list.scrollTo(0, 0);
    setPage(1);
    sort('newest');
    setSorted(true);
    setReviews(sortedReviews);
  };

  return (
    <StyledMain id="ratings-reviews" className="main-widget-container" ref={ref}>
      <StyledTitle id="inner-title">
        <div className="main-widget-title">
          RATINGS & REVIEWS
        </div>
      </StyledTitle>
      <StyledInner id="inner-main">
        <Ratings
          meta={meta}
          isLoaded={isLoaded}
          average={average}
          totalRatings={totalRatings}
          setRatingFilter={setRatingsFilter} // diff name from hook
          setSortOption={setSortOption}
          ratingFilter={ratingFilter}
        />
        {(filteredByRating) && (
          <Reviews
            productId={productId}
            totalRatings={totalRatings}
            reviews={filtered}
            moreReviews={moreReviews}
            setSort={sort}
            getReviews={getReviews}
            scrollMore={scrollMore}
            page={page}
            isLoaded={isLoaded}
            setPage={onFormSubmit}
          />
        )}
        {(!filteredByRating) && (
          <Reviews
            productId={productId}
            totalRatings={totalRatings}
            reviews={reviews}
            moreReviews={moreReviews}
            setSort={sort}
            getReviews={getReviews}
            scrollMore={scrollMore}
            page={page}
            isLoaded={isLoaded}
            setPage={onFormSubmit}
          />
        )}
      </StyledInner>
    </StyledMain>
  );
});

const StyledMain = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 70%;
  min-width: 500px;
  color: #1c1c1c;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
  border-bottom: .5px solid black;
`;
const StyledInner = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: top;
  width: 100%;
  padding-top: 1%;
  height: 100%
`;

const StyledTitle = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  font-size: 16px;
  font-weight: 400;
  margin-bottom: 1%;

`;

export default RatingsReviews;
