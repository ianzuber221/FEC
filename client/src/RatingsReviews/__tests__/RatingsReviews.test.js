import React from 'react';
import {
  render, screen, cleanup, fireEvent, waitFor, waitForElementToBeRemoved
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import RatingsReviews from '../RatingsReviews';
import { CurrentProductProvider } from '../../context';
import userEvent from '@testing-library/user-event';

require('dotenv').config();

test('should properly render out entire widget', () => {
  render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );

  const addbutton = screen.getByTestId('addbutton');
  expect(typeof addbutton).toEqual('object');
});

test('should get average rating upon load', async () => {
  render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  const totalratings = await screen.findByTestId('averagerating');
  //console.log(totalratings.textContent, 'is the freaking text content');
  expect(totalratings.textContent).not.toBe('undefined');
});

test('matches snapshot', async () => {
  const { asFragment } = await render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  // WAIT until element that has placeholder text to be removed
  await waitForElementToBeRemoved(screen.getByText('loading. . .'));

  expect(asFragment()).toMatchSnapshot(); // take snapshot
});


test('properly loads reviews', async () => {
  const { asFragment } = await render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  // WAIT until element that has placeholder text to be removed
  await waitForElementToBeRemoved(screen.getByText('loading reviewtiles'));
  const reviewtiles = screen.getByTestId('reviewtiles');
  expect(typeof reviewtiles).toEqual('object'); // take snapshot
});

// test('properly adds more reviews', async () => {
//   const { asFragment } = await render(
//     <CurrentProductProvider>
//       <RatingsReviews />
//     </CurrentProductProvider>
//   );
//   await waitForElementToBeRemoved(screen.getByText('loading reviewtiles'))
//   const reviewlist = await screen.findByTestId('reviewlist');
//   let before = reviewlist.textContent;
//   await console.log(reviewlist.textContent);
//   const morebutton = screen.getByTestId('morebutton');
//   await fireEvent.click(morebutton);
//   let after = reviewlist.textContent;
//   console.log(after);
//   expect(before).not.toBe(after); // compare
// });

test('properly pops up add review form', async () => {
  const { asFragment } = await render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  await waitForElementToBeRemoved(screen.getByText('loading reviewtiles'));
  const addbutton = screen.getByTestId('addbutton');
  fireEvent.click(addbutton);
  const addreviewform = screen.getByTestId('addreviewform');

  expect(typeof addreviewform).toEqual('object'); //form should exist in dom when addbutton is clicked.
})

test('shoul;d uh', async () => {
  const { asFragment } = await render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  await waitForElementToBeRemoved(screen.getByText('loading reviewtiles'));
})

test('shoul;d uh simulate select', async () => {
  const { asFragment } = await render(
    <CurrentProductProvider>
      <RatingsReviews />
    </CurrentProductProvider>
  );
  //const firstRender = asFragment();
  let newest = screen.findByTestId('newest-option');
  await userEvent.selectOptions(screen.getByTestId('sortoptions'), screen.getByTestId('newest-option'));
  expect(screen.getByTestId('newest-option').selected).toBeTruthy();
})