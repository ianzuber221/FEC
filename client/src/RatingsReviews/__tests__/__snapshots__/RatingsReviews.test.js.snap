// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gSAPjG GrqFE"
    id="ratings-reviews"
  >
    <div
      class="sc-iNWwEs hLYxht"
      id="inner-title"
    >
      <div>
        RATINGS & REVIEWS
      </div>
    </div>
    <div
      class="sc-lbxAil gylGro"
      id="inner-main"
    >
      <div
        class="sc-ksZaOG ewMooE"
        id="ratings-main"
      >
        <div>
          <div>
            <div
              class="sc-breuTD kCSoaS"
            >
              <div
                class="sc-evZas gwvPXL"
                data-testid="averagerating"
              >
                3.6
              </div>
              <div
                class="sc-kDDrLX irdQeg"
              >
                <div
                  class="sc-iqcoie jealnG"
                >
                  <div
                    class="sc-crXcEl bftmyc"
                    width="70%"
                  />
                </div>
              </div>
            </div>
            <div>
              <em>
                out of 144 reviews
              </em>
            </div>
          </div>
        </div>
        <div>
          <div
            class="sc-eCYdqJ jCNtKz"
          >
            <div>
              65% of reviewers recommend this item.
            </div>
            <div>
              <div
                class="sc-gsnTZi hfSvTy"
              >
                <div
                  class="sc-bczRLJ hfdcTo"
                >
                  5
                </div>
                <div
                  class="sc-dkzDqf kfoeuR"
                  id="bar-5"
                >
                  <div
                    class="sc-hKMtZM loSaXE"
                    width="34.02777777777778"
                  />
                </div>
              </div>
              <div
                class="sc-gsnTZi hfSvTy"
              >
                <div
                  class="sc-bczRLJ hfdcTo"
                >
                  4
                </div>
                <div
                  class="sc-dkzDqf kfoeuR"
                  id="bar-4"
                >
                  <div
                    class="sc-hKMtZM hKGPNK"
                    width="15.277777777777779"
                  />
                </div>
              </div>
              <div
                class="sc-gsnTZi hfSvTy"
              >
                <div
                  class="sc-bczRLJ hfdcTo"
                >
                  3
                </div>
                <div
                  class="sc-dkzDqf kfoeuR"
                  id="bar-3"
                >
                  <div
                    class="sc-hKMtZM hgUnSy"
                    width="33.33333333333333"
                  />
                </div>
              </div>
              <div
                class="sc-gsnTZi hfSvTy"
              >
                <div
                  class="sc-bczRLJ hfdcTo"
                >
                  2
                </div>
                <div
                  class="sc-dkzDqf kfoeuR"
                  id="bar-2"
                >
                  <div
                    class="sc-hKMtZM hMfuXF"
                    width="7.638888888888889"
                  />
                </div>
              </div>
              <div
                class="sc-gsnTZi hfSvTy"
              >
                <div
                  class="sc-bczRLJ hfdcTo"
                >
                  1
                </div>
                <div
                  class="sc-dkzDqf kfoeuR"
                  id="bar-1"
                >
                  <div
                    class="sc-hKMtZM hTGXkF"
                    width="9.722222222222223"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-jqUVSM ecRvcZ"
        >
          <div>
            <div
              class="sc-jSMfEi jXewUD"
            >
              <div
                class="sc-papXJ kmRgne"
              >
                Fit: 
              </div>
              <div
                class="sc-gKXOVf XPVAb"
              >
                <div
                  class="sc-iBkjds bUPpBi"
                  width="60%"
                />
              </div>
              <div
                class="sc-ftvSup fNURyZ"
              >
                <div>
                  small
                </div>
                <div>
                  perfect
                </div>
                <div>
                  large
                </div>
              </div>
            </div>
            <div
              class="sc-jSMfEi jXewUD"
            >
              <div
                class="sc-papXJ kmRgne"
              >
                Length: 
              </div>
              <div
                class="sc-gKXOVf XPVAb"
              >
                <div
                  class="sc-iBkjds frdGgF"
                  width="50%"
                />
              </div>
              <div
                class="sc-ftvSup fNURyZ"
              >
                <div>
                  short
                </div>
                <div>
                  perfect
                </div>
                <div>
                  long
                </div>
              </div>
            </div>
            <div
              class="sc-jSMfEi jXewUD"
            >
              <div
                class="sc-papXJ kmRgne"
              >
                Comfort: 
              </div>
              <div
                class="sc-gKXOVf XPVAb"
              >
                <div
                  class="sc-iBkjds frdGgF"
                  width="50%"
                />
              </div>
              <div
                class="sc-ftvSup fNURyZ"
              >
                <div>
                  poor
                </div>
                <div>
                  ok
                </div>
                <div>
                  perfect
                </div>
              </div>
            </div>
            <div
              class="sc-jSMfEi jXewUD"
            >
              <div
                class="sc-papXJ kmRgne"
              >
                Quality: 
              </div>
              <div
                class="sc-gKXOVf XPVAb"
              >
                <div
                  class="sc-iBkjds bUPpBi"
                  width="60%"
                />
              </div>
              <div
                class="sc-ftvSup fNURyZ"
              >
                <div>
                  poor
                </div>
                <div>
                  expected
                </div>
                <div>
                  perfect
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-jgbSNz bKYjiX"
        data-testid="reviews-1"
      >
        <div
          class="sc-kgUAyh iYsSbV"
          data-testid="sortbar-1"
        >
          5 of 144 reviews, sorted by
          <select
            class="sc-hTtwUo jMRut"
            data-testid="sortoptions"
            id="sort"
          >
            <option
              data-testid="relevance"
              value="relevance"
            >
              Relevance
            </option>
            <option
              data-testid="newest-option"
              value="newest"
            >
              Newest
            </option>
            <option
              value="helpful"
            >
              Helpfulness
            </option>
          </select>
        </div>
        <div
          class="sc-iIPllB fa-dgUB"
          data-testid="reviewlist"
        >
          <div
            data-testid="reviewtiles"
          >
            <div
              class="sc-kgflAQ gGRQQJ"
            >
              <div
                class="sc-fLlhyt etOmUv"
              >
                <div>
                  <div
                    class="sc-kDDrLX irdQeg"
                  >
                    <div
                      class="sc-iqcoie jealnG"
                    >
                      <div
                        class="sc-crXcEl loGwoC"
                        width="100%"
                      />
                    </div>
                  </div>
                </div>
                <div>
                  hello, July 18, 2022 
                </div>
              </div>
              <div
                class="sc-ivTmOn jrpZrq"
              >
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    best purchase ever
                  </h5>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <small>
                    don't report this one pleaseaaaa
                  </small>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    I recommend this product ✓
                  </h5>
                  <div
                    class="sc-dmRaPn lbvtwA"
                  >
                    <img
                      class="sc-llJcti cmdRRN"
                      data-testid="photo-0-0"
                      src="http://res.cloudinary.com/joehan/image/upload/v1658260875/b7xg9m7xthi6uzftq5me.jpg"
                    />
                    <img
                      class="sc-llJcti cmdRRN"
                      data-testid="photo-1-0"
                      src="http://res.cloudinary.com/joehan/image/upload/v1658260868/gcpjlscfmmcegdl8tdef.jpg"
                    />
                    <img
                      class="sc-llJcti cmdRRN"
                      data-testid="photo-2-0"
                      src="http://res.cloudinary.com/joehan/image/upload/v1658260853/svp4gsyb6pmu3qqhwgk0.jpg"
                    />
                    <img
                      class="sc-llJcti cmdRRN"
                      data-testid="photo-3-0"
                      src="http://res.cloudinary.com/joehan/image/upload/v1658260861/i1hfqtvyr0hkda2yq4vb.jpg"
                    />
                  </div>
                </div>
                <div
                  class="sc-bBrHrO gkbOEv"
                >
                  <span
                    data-testid="review-helpful"
                  >
                    Helpful?
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                    data-testid="helpful-button-0"
                  >
                    <u>
                      Yes
                    </u>
                  </span>
                  <span
                    data-testid="helpful-count-0"
                  >
                    (76)    |  
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                  >
                    <u>
                      Report
                    </u>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="sc-kgflAQ gGRQQJ"
            >
              <div
                class="sc-fLlhyt etOmUv"
              >
                <div>
                  <div
                    class="sc-kDDrLX irdQeg"
                  >
                    <div
                      class="sc-iqcoie jealnG"
                    >
                      <div
                        class="sc-crXcEl dbrdsl"
                        width="60%"
                      />
                    </div>
                  </div>
                </div>
                <div>
                  ReviewTileMan, July 16, 2022 
                </div>
              </div>
              <div
                class="sc-ivTmOn jrpZrq"
              >
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    New Review...I want to see this in my review tiles.
                  </h5>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <small>
                    Please work. Please populate in my review tiles. That would be nice pelase and thank you!
                  </small>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    I recommend this product ✓
                  </h5>
                </div>
                <div
                  class="sc-bBrHrO gkbOEv"
                >
                  <span
                    data-testid="review-helpful"
                  >
                    Helpful?
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                    data-testid="helpful-button-1"
                  >
                    <u>
                      Yes
                    </u>
                  </span>
                  <span
                    data-testid="helpful-count-1"
                  >
                    (11)    |  
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                  >
                    <u>
                      Report
                    </u>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="sc-kgflAQ gGRQQJ"
            >
              <div
                class="sc-fLlhyt etOmUv"
              >
                <div>
                  <div
                    class="sc-kDDrLX irdQeg"
                  >
                    <div
                      class="sc-iqcoie jealnG"
                    >
                      <div
                        class="sc-crXcEl eqLREF"
                        width="20%"
                      />
                    </div>
                  </div>
                </div>
                <div>
                  nopantsman, January 04, 2022 
                </div>
              </div>
              <div
                class="sc-ivTmOn jrpZrq"
              >
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    I HATE pants
                  </h5>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <small>
                    pants are the worst never ever get pants pants are bad wear shorts
                  </small>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                />
                <div
                  class="sc-bBrHrO gkbOEv"
                >
                  <span
                    data-testid="review-helpful"
                  >
                    Helpful?
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                    data-testid="helpful-button-2"
                  >
                    <u>
                      Yes
                    </u>
                  </span>
                  <span
                    data-testid="helpful-count-2"
                  >
                    (8)    |  
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                  >
                    <u>
                      Report
                    </u>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="sc-kgflAQ gGRQQJ"
            >
              <div
                class="sc-fLlhyt etOmUv"
              >
                <div>
                  <div
                    class="sc-kDDrLX irdQeg"
                  >
                    <div
                      class="sc-iqcoie jealnG"
                    >
                      <div
                        class="sc-crXcEl loGwoC"
                        width="100%"
                      />
                    </div>
                  </div>
                </div>
                <div>
                  nicole, September 15, 2021 
                </div>
              </div>
              <div
                class="sc-ivTmOn jrpZrq"
              >
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    These jeans are true to size!!!!
                  </h5>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <small>
                    Right now I have a lot of baby weight to lose, and I’m sitting at a size 14 jeans. Lately I’ve been wearing sweatpants every day because my 14 jeans feel slightly tight and i’ve been addicted to an elastic waistband for the last year. When ordering these jeans, I read the reviews, looked at the size chart, and decided I’d probably need a 16. I don’t like my skinny jeans to be a second skin. I like them tight, but not where you have to do magic tricks to get in them. I was discouraged, thinking I had gone up a size, but I ordered both the 14 and 16 with the intent to return whichever one didn’t work. When they came, I hopefully tried on the 14 first and THEY FIT 🎉🌟😃! In fact, they look a little saggy. So I’d totally recommend ordering your normal size- they have a lot of stretch to them. And if you’re feeling brave, maybe even try the next size down 🤷🏻‍♀️
                  </small>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    I recommend this product ✓
                  </h5>
                </div>
                <div
                  class="sc-bBrHrO gkbOEv"
                >
                  <span
                    data-testid="review-helpful"
                  >
                    Helpful?
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                    data-testid="helpful-button-3"
                  >
                    <u>
                      Yes
                    </u>
                  </span>
                  <span
                    data-testid="helpful-count-3"
                  >
                    (4)    |  
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                  >
                    <u>
                      Report
                    </u>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="sc-kgflAQ gGRQQJ"
            >
              <div
                class="sc-fLlhyt etOmUv"
              >
                <div>
                  <div
                    class="sc-kDDrLX irdQeg"
                  >
                    <div
                      class="sc-iqcoie jealnG"
                    >
                      <div
                        class="sc-crXcEl loGwoC"
                        width="100%"
                      />
                    </div>
                  </div>
                </div>
                <div>
                  jose, September 15, 2021 
                </div>
              </div>
              <div
                class="sc-ivTmOn jrpZrq"
              >
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    Excellent pants
                  </h5>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <small>
                    me gusta estos pantelones mucho
                  </small>
                </div>
                <div
                  class="sc-idiyUo eWPMfl"
                >
                  <h5>
                    I recommend this product ✓
                  </h5>
                </div>
                <div
                  class="sc-bBrHrO gkbOEv"
                >
                  <span
                    data-testid="review-helpful"
                  >
                    Helpful?
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                    data-testid="helpful-button-4"
                  >
                    <u>
                      Yes
                    </u>
                  </span>
                  <span
                    data-testid="helpful-count-4"
                  >
                    (4)    |  
                  </span>
                  <span
                    class="sc-dIouRR jMDPQB"
                  >
                    <u>
                      Report
                    </u>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-cOFTSb dsxEYY"
          data-testid="addbar-1"
        >
          <button
            class="sc-BeQoi dtjXfM"
            data-testid="morebutton"
          >
            MORE REVIEWS
          </button>
          <button
            class="sc-BeQoi dtjXfM"
            data-testid="addbutton"
          >
            ADD A REVIEW +
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
